File: algorithms/sorting/cpp/MergeSort.cpp
#include <vector>
using namespace std;

class Sorter {
private:
  vector<int> v;

public:
  Sorter(vector<int> &v) : v(v) {}

  void merge(vector<int> &v, int l, int r, int m) {
    vector<int> temp;

    int left = l;
    int right = m + 1;

    while (left <= m && right <= r) {
      if (v[left] < v[right]) {
        temp.push_back(v[left]);
        left++;
      } else {
        temp.push_back(v[right]);
        right++;
      }
    }

    while (left <= m) {
      temp.push_back(v[left]);
      left++;
    }

    while (right <= r) {
      temp.push_back(v[right]);
      right++;
    }

    for (int i = l; i <= r; i++) {
      v[i] = temp[i - l];
    }
  }

  void mergeSort(vector<int> &v, int l, int r) {
    if (l < r) {
      int mid = (l + r) / 2;

      mergeSort(v, l, mid);
      mergeSort(v, mid + 1, r);

      merge(v, l, r, mid);
    }
  }
};

File: algorithms/sorting/go/mergesort.go
package main

import "fmt"

func mergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	// divide
	mid := len(arr) / 2
	left := mergeSort(arr[:mid])
	right := mergeSort(arr[mid:])

	// conquer
	return merge(left, right)
}

func merge(left, right []int) []int {
	result := []int{}
	i, j := 0, 0

	// Merge two sorted slices
	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	// Append remaining elements
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)

	return result
}

func main() {
	arr := []int{4, 1, 47, 64, 23, 6, 37,69}

	sorted := mergeSort(arr)
	fmt.Println("Sorted array: ", sorted)
}
